src/cryptonote_core/service_node_list.cpp

```diff
@@ -1236,99 +1325,180 @@ namespace service_nodes
+          pulse_candidates.reserve(active_snode_list.size());
 
-      for (; i < num_validators + num_workers; i++)
-      {
-        size_t j = pub_keys_indexes[i];
-        if (j < active_snode_list.size())
-          quorum->workers.push_back(active_snode_list[j].first);
-        else
-          quorum->workers.push_back(decomm_snode_list[j - active_snode_list.size()].first);
+          // TODO(doyle): Support leaders failing and iterating the round for this block.
+          // NOTE: Find the leader for this block
+          {
+            auto best_leader_index = static_cast<size_t>(-1);
+            auto best_reward_time  = std::make_pair(static_cast<uint64_t>(-1), static_cast<uint32_t>(-1));
+            for (size_t node_index = 0; node_index < active_snode_list.size(); node_index++)
+            {
+              pubkey_and_sninfo const &node = active_snode_list[node_index];
+              auto reward_time = std::make_pair(node.second->last_reward_block_height, node.second->last_reward_transaction_index);
+              if (reward_time < best_reward_time)
+              {
+                best_reward_time  = reward_time;
+                best_leader_index = node_index;
+              }
+
+              pulse_candidates.push_back(&node);
+            }
+
+            quorum->workers.push_back(pulse_candidates[best_leader_index]->first);
+            pulse_candidates.erase(pulse_candidates.begin() + best_leader_index);
+          }
+
+          // NOTE: Sort ascending in height i.e. sort preferring the longest time since the validator was in a Pulse quorum.
+          std::stable_sort(pulse_candidates.begin(), pulse_candidates.end(), [](pubkey_and_sninfo const *a, pubkey_and_sninfo const *b) {
+                      auto a_key  = std::make_pair(a->second->pulse_sort_key.last_height_validating_in_quorum, a->second->pulse_sort_key.quorum_index);
+                      auto b_key  = std::make_pair(b->second->pulse_sort_key.last_height_validating_in_quorum, b->second->pulse_sort_key.quorum_index);
+                      bool result = a_key < b_key;
+                      return result;
+                    });
+
+          // NOTE: Divide the list in half, select validators from the first half of the list.
+          // Remove them from the list of candidates once chosen.
+          size_t const midpoint = pulse_candidates.size() / 2;
+          for (size_t i = 0; i < pulse_entropy.size(); i++)
+          {
+            size_t const partition_index = std::min(midpoint, pulse_candidates.size());
+            std::mt19937_64 rng(quorum_rng_seed(pulse_entropy[i], type));
+            size_t candidate_index             = tools::uniform_distribution_portable(rng, partition_index);
+            pubkey_and_sninfo const *candidate = pulse_candidates[candidate_index];
+
+            quorum->validators.push_back(candidate->first);
+            pulse_candidates.erase(pulse_candidates.begin() + candidate_index);
```

Comment:
This `erase` is going to be pretty heavy (each call is O(N)).  You can avoid it by doing something like:

- start out (before the loop) with a `auto begin = pulse_candidates.begin();`
- move partition_index outside the loop
- loop:
  - change the random count used in `candidate_index` to subtract how far we've come: `partition_index - std::distance(pulse_candidates.begin(), begin)`
  - swap the element at `begin` with the element at `begin + partition_index`
  - increment `begin`

Then at the end of the loop, `[pulse_candidates.begin, begin)` is the selection, so `pulse_candidates.erase(begin, pulse_candidates.end())`, and now you've got `pulse_candidates` set to the selected elements.  (You can then loop through the truncated list to populate quorum validators and update selected candidate sort keys).